WireAPI/Sources/WireAPI/Authentication/AuthenticationManager.swift

```diff
@@ -0,0 +1,156 @@
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireFoundation
+
+// sourcery: AutoMockable
+protocol AuthenticationManagerProtocol {
+
+    func getValidAccessToken() async throws -> AccessToken
+    func refreshAccessToken() async throws -> AccessToken
+
+}
+
+actor AuthenticationManager: AuthenticationManagerProtocol {
+
+    private var currentToken: CurrentToken?
+    private let clientID: String
+    private let cookieStorage: CookieStorage
+    private let networkService: NetworkService
+
+    init(
+        clientID: String,
+        cookieStorage: CookieStorage,
+        networkService: NetworkService
+    ) {
+        self.clientID = clientID
+        self.cookieStorage = cookieStorage
+        self.networkService = networkService
+    }
+
+    func getValidAccessToken() async throws -> AccessToken {
+        switch currentToken {
+        case .renewing(let task):
+            return try await task.value
+
+        case .cached(let accessToken) where !accessToken.isExpiring:
+            return accessToken
+
+        default:
+            return try await refreshAccessToken()
+        }
+    }
+
+    func refreshAccessToken() async throws -> AccessToken {
+        switch currentToken {
+        case .renewing(let task):
+            // A request is in flight, we wait for its result.
+            return try await task.value
+
+        case .cached(let accessToken):
+            // We are renewing a token.
+            let task = makeRenewTokenTask(lastKnownAccessToken: accessToken)
+            currentToken = .renewing(task)
+            return try await task.value
+
+        default:
+            // We are getting the first token.
+            let task = makeRenewTokenTask(lastKnownAccessToken: nil)
+            currentToken = .renewing(task)
+            return try await task.value
+        }
+    }
+
+    private func makeRenewTokenTask(
+        lastKnownAccessToken: AccessToken?
+    ) -> Task<AccessToken, any Error> {
+        Task {
+            let cookies = try await cookieStorage.fetchCookies()
+
+            var request = try URLRequestBuilder(path: "/access")
+                .withQueryItem(name: "client_id", value: clientID)
+                .withMethod(.post)
+                .withAcceptType(.json)
+                .withCookies(cookies)
+                .build()
+
+            if let lastKnownAccessToken {
+                request.setAccessToken(lastKnownAccessToken)
+            }
+
+            let (data, response) = try await networkService.executeRequest(request)
+
+            let accessToken = try ResponseParser()
+                .success(code: .ok, type: AccessTokenPayload.self)
+                .failure(code: .forbidden, label: "invalid-credentials", error: APIServiceError.invalidCredentials)
+                .parse(code: response.statusCode, data: data)
+
+            currentToken = .cached(accessToken)
```

Comment:
For symmetry I think this should be set by the caller (refreshAccessToken()). And if this method throws current token can be set to `nil` which I expect is the correct behavior.

As it stands I think there is a bug:

If this exits due to failure, then we will be in the renewing case and any future call to getValidAccessToken() will await indefinitely (I think).